### PR TITLE
HitShape, query trait via actor cached targetable positions.

### PR DESCRIPTION
**Pull request by anvilvapre**
HitShape is the most looked up trait per per actor using `TraitsImplementing`. 

Using this method will search the global `TraitDictionary` for all actors with the trait and after search for the actor on which the method was called. 

Changes:
- Rather than using `TraitsImplementing` use the already available `enabledTargetablePositions` to find `HitShape` traits. 
- Add `Actor.GetEnabledTargetablePositions`.
- Because of this some Linq statements rewritten to foreach loops. 

Optional suggestions:
- Optionally rename `Actor.GetTargetablePositions` to `Actor.GetEnabledTargetableWorldPositions`
-  Move IHitShape to OpenRA.Game `TraitInterfaces`.  Add `ITargertablePositionsInterface.TargetableShapes()` to return an enumeration of HitShapes.

Assumptions:
- Casting `ITargetablePositions` to `HitShape` is faster than each time searching for an actor in the trait container.

Advantage:
- Save 300k times searching for a actor in the trait container. 

Statistics on traits being most often queried for an actor using `TraitDictionary.WithInterface`. Shelllmap + a short bot game,.
````
    391 OpenRA.Mods.Common.Traits.IGainsExperienceModifier
    429 OpenRA.Mods.Common.Traits.INotifyProduction
    458 OpenRA.Mods.Common.Traits.Tooltip
    607 OpenRA.Mods.Common.Traits.ISpeedModifier
    719 OpenRA.Mods.Common.Traits.ExternalCondition
    764 OpenRA.Mods.Common.Traits.Locomotor
    932 OpenRA.Mods.Common.Traits.IWallConnector
   1225 OpenRA.Mods.Common.Traits.SquadManagerBotModule
   1355 OpenRA.Mods.Common.Traits.Exit
   1384 OpenRA.Mods.Common.Traits.INotifyDamage
   1384 OpenRA.Mods.Common.Traits.INotifyKilled
   1400 OpenRA.Mods.Common.Traits.INotifyDocking
   1487 OpenRA.Mods.Common.Traits.SupportPower
   1489 OpenRA.Traits.INotifyAddedToWorld
   1661 OpenRA.Mods.Common.Traits.IDamageModifier
   1852 OpenRA.Mods.Common.Traits.Power
   1980 OpenRA.Traits.IProvideTooltipInfo
   1980 OpenRA.Traits.ITooltip
   2308 OpenRA.Mods.Common.Traits.AttackFrontal
   2363 OpenRA.Mods.Common.Traits.GivesBuildableArea
   2819 OpenRA.Traits.ISelectionBar
   6124 OpenRA.Mods.Common.Traits.SmudgeLayer
   6607 OpenRA.Mods.Common.Traits.AttackBase
   6763 OpenRA.Mods.Common.Traits.INotifyHarvesterAction
   8088 OpenRA.Mods.Common.Traits.RevealsShroud
   8374 OpenRA.Mods.Common.Traits.Armament
  16677 OpenRA.Mods.Common.Traits.Armor
  18434 OpenRA.Traits.IVisibilityModifier
  20703 OpenRA.Mods.Common.Traits.IBlocksProjectiles
  26319 OpenRA.Traits.IIssueOrder
  27066 OpenRA.Traits.IRenderAboveShroudWhenSelected
  27066 OpenRA.Traits.IRenderAnnotationsWhenSelected
  46052 OpenRA.Mods.Common.Traits.INotifyAppliedDamage
  57296 (not
  64625 OpenRA.Mods.Common.Traits.RejectsOrders
  77929 OpenRA.Traits.IRenderOverlay
  77931 OpenRA.Traits.IPaletteModifier
  93490 OpenRA.Mods.Common.Traits.IDisableEnemyAutoTarget
  98184 OpenRA.Mods.Common.Traits.ICrushable
 185125 OpenRA.Mods.Common.Traits.INotifyBlockingMove
 324904 OpenRA.Mods.Common.Traits.HitShape
````

### CODE REVIEW

File: OpenRA.Mods.Common/Warheads/DamageWarhead.cs [hunk 1]
```diff
@@ -50,8 +50,20 @@ public override void DoImpact(in Target target, WarheadArgs args)
 				if (!IsValidAgainst(victim, firedBy))
 					return;
 
-				var closestActiveShape = victim.TraitsImplementing<HitShape>().Where(Exts.IsTraitEnabled)
-					.MinByOrDefault(t => t.DistanceFromEdge(victim, victim.CenterPosition));
+				HitShape closestActiveShape = null;
+				var closestDistance = new WDist(int.MaxValue);
+				foreach (var targetPos in victim.GetEnabledTargetablePositions())
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/Warheads/DamageWarhead.cs]: ~~What is the problem with keeping this as LINQ statement?~~
```
var closestActiveShape = victim.GetEnabledTargetablePositions().MinByOrDefault(t => t.DistanceFromEdge(victim, victim.CenterPosition));
```
Edit: Never mind, I forgot about the cast. We could still do that in LINQ but I'm not sure if it's better.
```
var closestActiveShape = victim.GetEnabledTargetablePositions().MinByOrDefault(t =>
{
	if (t is HitShape h)
		return h.DistanceFromEdge(victim, victim.CenterPosition));
	else
		return WDist.MaxValue;
});
```

Comment (abcdefg30) [reply OpenRA.Mods.Common/Warheads/DamageWarhead.cs]: Actually I think I like the LINQ version more since it is less verbose.

Comment (anvilvapre) [reply OpenRA.Mods.Common/Warheads/DamageWarhead.cs]: I wanted to use your example. However i need both the hit shape and distance. As the hitshape is passed below.

Comment (abcdefg30) [reply OpenRA.Mods.Common/Warheads/DamageWarhead.cs]: That should return the closest hitshape. The distance shouldn't be needed below?

Comment (anvilvapre) [reply OpenRA.Mods.Common/Warheads/DamageWarhead.cs]: I looked at the `SpreadDamageWarHead`. There both are used. 

I changed it for this file.

File: OpenRA.Game/Actor.cs [hunk 2]
```diff
@@ -536,6 +536,11 @@ public IEnumerable<WPos> GetTargetablePositions()
 			return new[] { CenterPosition };
 		}
 
+		public IEnumerable<ITargetablePositions> GetEnabledTargetablePositions()
+		{
+			return enabledTargetablePositions;
+		}
+
```
Comment (PunkPun) [inline OpenRA.Game/Actor.cs]: Couldn't we make `enabledTargetablePositions` private set public get instead?

File: OpenRA.Mods.Common/Warheads/DamageWarhead.cs [hunk 3]
```diff
@@ -50,14 +50,19 @@ public override void DoImpact(in Target target, WarheadArgs args)
 				if (!IsValidAgainst(victim, firedBy))
 					return;
 
-				var closestActiveShape = victim.TraitsImplementing<HitShape>().Where(Exts.IsTraitEnabled)
-					.MinByOrDefault(t => t.DistanceFromEdge(victim, victim.CenterPosition));
+				var closestActiveShape = victim.GetEnabledTargetablePositions().MinByOrDefault(t =>
```
Comment (PunkPun) [inline OpenRA.Mods.Common/Warheads/DamageWarhead.cs]: IMO it's better to move cast to `HitShape` here

File: OpenRA.Mods.Common/Warheads/SpreadDamageWarhead.cs [hunk 4]
```diff
@@ -56,26 +56,35 @@ protected override void DoImpact(WPos pos, Actor firedBy, WarheadArgs args)
 			var debugVis = firedBy.World.WorldActor.TraitOrDefault<DebugVisualizations>();
 			if (debugVis != null && debugVis.CombatGeometry)
 				firedBy.World.WorldActor.Trait<WarheadDebugOverlay>().AddImpact(pos, effectiveRange, DebugOverlayColor);
-
```
Comment (PunkPun) [inline OpenRA.Mods.Common/Warheads/SpreadDamageWarhead.cs]: Revert this change, removing this white space violates our style conventions

File: OpenRA.Mods.Common/Warheads/TargetDamageWarhead.cs [hunk 5]
```diff
@@ -35,20 +34,30 @@ protected override void DoImpact(WPos pos, Actor firedBy, WarheadArgs args)
...
-					.Where(Exts.IsTraitEnabled)
-					.Select(s => (HitShape: s, Distance: s.DistanceFromEdge(victim, pos)))
-					.MinByOrDefault(s => s.Distance);
+				HitShape closestActiveShape = null;
+				var closestDistance = new WDist(int.MaxValue);
+				foreach (var targetPos in victim.GetEnabledTargetablePositions())
+				{
+					if (targetPos is HitShape h)
+					{
+						var distance = h.DistanceFromEdge(victim, pos);
```
Comment (PunkPun) [inline OpenRA.Mods.Common/Warheads/TargetDamageWarhead.cs]: wouldn't it be better to use ints? That way we wouldn't need to use `new WDist(int.MaxValue)`
```suggestion
						var distance = h.DistanceFromEdge(victim, pos).Length;
```

File: OpenRA.Game/Actor.cs (hunk 2 above)
Comment (PunkPun) [reply OpenRA.Game/Actor.cs:543]: Also I don't think this array is dynamic. I don't see code that would check for `HitShape` becoming disabled, or new HitShapes becoming  enabled

File: OpenRA.Mods.Common/Warheads/SpreadDamageWarhead.cs [hunk 6]
```diff
@@ -56,26 +56,35 @@ protected override void DoImpact(WPos pos, Actor firedBy, WarheadArgs args)
...
 			{
 				if (!IsValidAgainst(victim, firedBy))
 					continue;
 
-				var closestActiveShape = victim.TraitsImplementing<HitShape>()
-					.Where(Exts.IsTraitEnabled)
-					.Select(s => (HitShape: s, Distance: s.DistanceFromEdge(victim, pos)))
-					.MinByOrDefault(s => s.Distance);
+				HitShape closestActiveShape = null;
+				var closestDistance = new WDist(int.MaxValue);
```
Comment (PunkPun) [inline OpenRA.Mods.Common/Warheads/SpreadDamageWarhead.cs]: same here

**Review by PunkPun — COMMENTED** [review]
I think the `// PERF` comment should be added everywhere we cast to HitShape, not just the few instances

File: OpenRA.Game/Actor.cs (hunk 2 above)
Comment (PunkPun) [reply OpenRA.Game/Actor.cs:543]: We should rather store all `ITargetablePositions` instead and in functions provide enabled ones

**Comment by anvilvapre** [comment]
> I think the `// PERF` comment should be added everywhere we cast to HitShape, not just the few instances

How about removing it everywhere? Isn't just just common to use a local cache trait over looking it up in the trait container?

File: OpenRA.Game/Actor.cs (hunk 2 above)
Comment (anvilvapre) [reply OpenRA.Game/Actor.cs]: Sure. I went with Get because `enabledTargetablePositions` exists, which returns a center position.

File: OpenRA.Mods.Common/Warheads/SpreadDamageWarhead.cs (hunk 4 above)
Comment (anvilvapre) [reply OpenRA.Mods.Common/Warheads/SpreadDamageWarhead.cs]: Note that `make check` did not complain.

**Comment by anvilvapre** [comment]
> I think the `// PERF` comment should be added everywhere we cast to HitShape, not just the few instances

Done.

**Review by PunkPun — DISMISSED** [review]
Untested, but code LGTM

**Comment by penev92** [comment]
Reading around your changes, I noticed some minor housekeeping is in order. I feel it would be best done *before* the changes in this PR, but since it's small enough that could also be done by you cherry-picking my commit and rebasing yours on top of it.
Or if people would prefer a separate PR I can open one.
The idea is that a few redundant things are removed and you even get our public getter.
https://github.com/OpenRA/OpenRA/compare/bleed...penev92:OpenRA:targetablePositionsCleanup

*P.S.: Sorry, I see in the mean time you updated the PR to be closer to what I have, but I still think mine is even cleaner, so my request stands.*

**Comment by RoosterDragon** [comment]
The reason to cache an `IEnumerable` such as `items.Where(x => x.Foo)` is to avoid repeating the allocations that occur for the lambda and the `Where` machinery. Usually such allocations aren't worth worrying about. But the various helper methods on Actor are called often so it can be worth bearing in mind.

**Comment by penev92** [comment]
Never mind my comment then. This does call for some explanatory comments though, but that's not related to this PR.

File: OpenRA.Mods.Common/Warheads/TargetDamageWarhead.cs [hunk 7]
```diff
@@ -36,20 +35,32 @@ protected override void DoImpact(WPos pos, Actor firedBy, WarheadArgs args)
...
-					.Where(Exts.IsTraitEnabled)
-					.Select(s => (HitShape: s, Distance: s.DistanceFromEdge(victim, pos)))
-					.MinByOrDefault(s => s.Distance);
+				HitShape closestActiveShape = null;
+				var closestDistance = int.MaxValue;
+
+				// PERF: Avoid using TraitsImplementing<HitShape> that needs to find the actor in the trait dictionary.
+				foreach (var targetPos in victim.EnabledTargetablePositions)
+				{
+					if (targetPos is HitShape h)
```
Comment (Mailaender) [inline OpenRA.Mods.Common/Warheads/TargetDamageWarhead.cs]:
```suggestion
					if (targetPos is HitShape hitshape)
```
maybe? one letter variables are cryptic.

File: OpenRA.Mods.Common/WorldExtensions.cs [hunk 8]
```diff
@@ -49,9 +49,11 @@ public static IEnumerable<Actor> FindActorsOnLine(this World world, WPos lineSta
 			foreach (var currActor in actorsInSquare)
 			{
 				var actorWidth = 0;
-				var shapes = currActor.TraitsImplementing<HitShape>().Where(Exts.IsTraitEnabled);
-				if (shapes.Any())
-					actorWidth = shapes.Max(h => h.Info.Type.OuterRadius.Length);
+
+				// PERF: Avoid using TraitsImplementing<HitShape> that needs to find the actor in the trait dictionary.
+				foreach (var targetPos in currActor.EnabledTargetablePositions)
+					if (targetPos is HitShape h)
```
Comment (Mailaender) [inline OpenRA.Mods.Common/WorldExtensions.cs]:
```suggestion
					if (targetPos is HitShape hitshape)
```
same here, this isn't a LINQ lambda anymore.

**Review by Mailaender — DISMISSED** [review]
Hitshapes still function in-game.

**Comment by Mailaender** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/cff4af493a04a3ee0efd4d794f1f0063e3c1c296...487f37c0fb5536899cee2633530c31d8589dab93)
